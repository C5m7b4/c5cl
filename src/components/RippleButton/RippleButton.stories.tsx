import React from 'react';
import { Story, Meta } from '@storybook/react';

import RippleButton, { RippleButtonProps } from './RippleButton';

export default {
  title: 'Buttons',
  component: RippleButton,
} as Meta;

const Template: Story<RippleButtonProps> = (args) => <RippleButton {...args} />;

export const RippleButtonExample = Template.bind({});
RippleButtonExample.args = {
  text: 'Click Me',
};

export const SuccessRippleButtonExample = Template.bind({});
SuccessRippleButtonExample.args = {
  text: 'Success',
  type: 'success',
};

export const NormalRippleButtonExample = Template.bind({});
NormalRippleButtonExample.args = {
  text: 'Normal',
  type: 'normal',
};

export const GreenRippleButtonExample = Template.bind({});
GreenRippleButtonExample.args = {
  text: 'Green',
  type: 'green',
};

export const PurpleRippleButtonExample = Template.bind({});
PurpleRippleButtonExample.args = {
  text: 'Purple',
  type: 'purple',
};

export const DangerRippleButtonExample = Template.bind({});
DangerRippleButtonExample.args = {
  text: 'Danger',
  type: 'danger',
};

export const InfoRippleButtonExample = Template.bind({});
InfoRippleButtonExample.args = {
  text: 'Info',
  type: 'info',
};

export const ErrorRippleButtonExample = Template.bind({});
ErrorRippleButtonExample.args = {
  text: 'Error',
  type: 'error',
};
