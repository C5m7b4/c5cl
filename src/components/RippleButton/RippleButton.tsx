import React, { useEffect, useRef } from 'react';
import { NormalButtonType } from '../Button/Button';
import './RippleButton.css';

export interface RippleButtonProps {
  text: string;
  onClick: (e: React.MouseEvent) => void;
  color?: string;
  hoverColor?: string;
  textColor?: string;
  style?: React.CSSProperties;
  type?: NormalButtonType;
}

function RippleButton(props: RippleButtonProps) {
  const btn = useRef<HTMLButtonElement>(null);
  let {
    color = '#2952e3',
    hoverColor = '#2546bd',
    textColor = '#fff',
    type = 'normal',
  } = props;

  switch (type) {
    case 'green':
      color = 'limegreen';
      hoverColor = '#2ea62e';
      textColor = '#fff';
      break;
    case 'purple':
      color = '#ba1eba';
      hoverColor = '#9c229c';
      textColor = '#fff';
      break;
    case 'danger':
      color = '#e8bb25';
      hoverColor = '#c9a734';
      textColor = "#fff';";
      break;
    case 'info':
      color = '#5a95cd';
      hoverColor = '#5680a8';
      textColor = '#fff';
      break;
    case 'success':
      color = '#1bcc4a';
      hoverColor = 'rgb(35 169 71)';
      textColor = '#fff';
      break;
    case 'error':
      color = 'red';
      hoverColor = '#bf2424';
      textColor = '#fff';
      break;
    default:
    // do nothing;
  }

  const mouseEnter = (e: React.MouseEvent) => {
    // @ts-ignore
    e.target.style.backgroundColor = hoverColor;
  };

  const mouseLeave = (e: React.MouseEvent) => {
    // @ts-ignore
    e.target.style.backgroundColor = color;
  };

  const click = (e: React.MouseEvent) => {
    const x = e.clientX;
    const y = e.clientY;

    // @ts-ignore
    const buttonTop = e.target?.offsetTop;
    // @ts-ignore
    const buttonLeft = e.target?.offsetLeft;

    const xInside = x - buttonLeft;
    const yInside = y - buttonTop;

    const circle = document.createElement('span');
    circle.classList.add('ripple-button-circle');

    circle.style.top = yInside + 'px';
    circle.style.left = xInside + 'px';

    btn.current ? btn.current.appendChild(circle) : null;

    props.onClick(e);

    setTimeout(() => {
      circle.remove();
    }, 500);
  };

  useEffect(() => {
    /* istanbul ignore else */
    if (btn.current) {
      btn.current.style.backgroundColor = color;
      btn.current.style.color = textColor;
    }
  }, []);

  return (
    <button
      ref={btn}
      style={props.style}
      className="c5-ripple-button"
      onClick={click}
      onMouseEnter={mouseEnter}
      onMouseLeave={mouseLeave}
    >
      {props.text}
      <span className="ripple-button-circle"></span>
    </button>
  );
}

export default RippleButton;
