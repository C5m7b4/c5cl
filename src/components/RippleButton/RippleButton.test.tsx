import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import fireEvent from '@testing-library/user-event';

import RippleButton from './RippleButton';

// afterEach(cleanup);

// beforeAll(() =>
//   jest.spyOn(React, 'useEffect').mockImplementation(React.useLayoutEffect)
// );
// // @ts-ignore
// afterAll(() => React.useEffect.mockRestore());

describe('RippleButton', () => {
  test('should render correctly', () => {
    const testFn = jest.fn();
    const frag = render(<RippleButton text="Click Me" onClick={testFn} />);

    expect(frag).toMatchSnapshot();
  });
  test('should handle click event', () => {
    jest.useFakeTimers('modern');
    const testFn = jest.fn();
    render(<RippleButton text="Click Me" onClick={testFn} />);

    fireEvent.click(screen.getByText('Click Me'));
  });
  test('should simulate a mouse leave event', () => {
    const testFn = jest.fn();
    render(<RippleButton text="Click Me" onClick={testFn} />);

    fireEvent.hover(screen.getByText('Click Me'));
    fireEvent.unhover(screen.getByText('Click Me'));
  });
  test('should unmount', () => {
    jest.useFakeTimers();
    const testFn = jest.fn();
    const { container, unmount, rerender } = render(
      <RippleButton text="Click Me" onClick={testFn} />
    );
    rerender(<RippleButton text="Click Me again" onClick={testFn} />);
    unmount();
    jest.advanceTimersByTime(500);
  });
  test('should handle green type', () => {
    const testFn = jest.fn();
    const { container, unmount, rerender } = render(
      <RippleButton text="Click Me" onClick={testFn} type="green" />
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle purple type', () => {
    const testFn = jest.fn();
    const { container, unmount, rerender } = render(
      <RippleButton text="Click Me" onClick={testFn} type="purple" />
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle danger type', () => {
    const testFn = jest.fn();
    const { container, unmount, rerender } = render(
      <RippleButton text="Click Me" onClick={testFn} type="danger" />
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle info type', () => {
    const testFn = jest.fn();
    const { container, unmount, rerender } = render(
      <RippleButton text="Click Me" onClick={testFn} type="info" />
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle success type', () => {
    const testFn = jest.fn();
    const { container, unmount, rerender } = render(
      <RippleButton text="Click Me" onClick={testFn} type="success" />
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle error type', () => {
    const testFn = jest.fn();
    const { container, unmount, rerender } = render(
      <RippleButton text="Click Me" onClick={testFn} type="error" />
    );
    expect(container).toMatchSnapshot();
  });
});
