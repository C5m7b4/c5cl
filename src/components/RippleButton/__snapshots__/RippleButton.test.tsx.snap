// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RippleButton should handle danger type 1`] = `
<div>
  <button
    class="c5-ripple-button"
    style="background-color: rgb(232, 187, 37);"
  >
    Click Me
    <span
      class="ripple-button-circle"
    />
  </button>
</div>
`;

exports[`RippleButton should handle error type 1`] = `
<div>
  <button
    class="c5-ripple-button"
    style="background-color: red; color: rgb(255, 255, 255);"
  >
    Click Me
    <span
      class="ripple-button-circle"
    />
  </button>
</div>
`;

exports[`RippleButton should handle green type 1`] = `
<div>
  <button
    class="c5-ripple-button"
    style="background-color: limegreen; color: rgb(255, 255, 255);"
  >
    Click Me
    <span
      class="ripple-button-circle"
    />
  </button>
</div>
`;

exports[`RippleButton should handle info type 1`] = `
<div>
  <button
    class="c5-ripple-button"
    style="background-color: rgb(90, 149, 205); color: rgb(255, 255, 255);"
  >
    Click Me
    <span
      class="ripple-button-circle"
    />
  </button>
</div>
`;

exports[`RippleButton should handle purple type 1`] = `
<div>
  <button
    class="c5-ripple-button"
    style="background-color: rgb(186, 30, 186); color: rgb(255, 255, 255);"
  >
    Click Me
    <span
      class="ripple-button-circle"
    />
  </button>
</div>
`;

exports[`RippleButton should handle success type 1`] = `
<div>
  <button
    class="c5-ripple-button"
    style="background-color: rgb(27, 204, 74); color: rgb(255, 255, 255);"
  >
    Click Me
    <span
      class="ripple-button-circle"
    />
  </button>
</div>
`;

exports[`RippleButton should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="c5-ripple-button"
        style="background-color: rgb(41, 82, 227); color: rgb(255, 255, 255);"
      >
        Click Me
        <span
          class="ripple-button-circle"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="c5-ripple-button"
      style="background-color: rgb(41, 82, 227); color: rgb(255, 255, 255);"
    >
      Click Me
      <span
        class="ripple-button-circle"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
